clean up deployment script for ubuntu

dalanlan opened the pull request:
- Fix #17014.
- Carry clean up work for messy script for deployment on ubuntu. Ref #15060

dalanlan committed:
clean up deployment script for ubuntu

dalanlan commented:
@MikeSpreitzer It did have pb (Carry #15060 w/o second thought) and i realized that as well. Yet it's not finished and i would ping you once it's done so that you can review it later;-)

WIZARD-CXY commented on cluster/ubuntu/reconfDocker.sh:
```diff
@@ -28,7 +28,7 @@ function config_etcd {
   attempt=0
   while true; do
-    /opt/bin/etcdctl get /coreos.com/network/config
+    /opt/bin/etcdctl --no-sync get /coreos.com/network/config
```
Is it a new etcd feature??

dalanlan replied:
sort of

dalanlan replied:
Ref https://github.com/coreos/etcd/blob/master/etcdctl/command/util.go#L210~L218

dalanlan commented:
@zmerlynn @MikeSpreitzer Please kindly review this once you're available

dalanlan committed:
fix messy quote thing

dalanlan commented on cluster/ubuntu/util.sh:
```diff
@@ -346,99 +349,229 @@ function kube-up() {
+      
+      groupadd -f -r kube-cert
+      "${PROXY_SETTING}" ~/kube/make-ca-cert.sh "${MASTER_IP}" "${EXTRA_SANS}"
+      mkdir -p /opt/bin/
```
@MikeSpreitzer Please check this 'cause i could barely catch what `squote` exactly means :-)

MikeSpreitzer replied:
squote is hacker jargon for Single-QUOTE.  dquote is Double-QUOTE.

MikeSpreitzer commented on cluster/ubuntu/util.sh:
```diff
@@ -346,99 +349,229 @@ function kube-up() {
+      cp ~/kube/init_scripts/* /etc/init.d/
+      
+      groupadd -f -r kube-cert
+      "${PROXY_SETTING}" ~/kube/make-ca-cert.sh "${MASTER_IP}" "${EXTRA_SANS}"
```
Please note that three levels of quoting are being attempted here, but it is not done correctly.
The intended outermost level of quoting begins on line 379 and ends on line 402; this is constructing a long string argument to and SSH command to ${MASTER}.  The intended middle level of quoting begins on line 391 and ends on line 402; it is constructing a multi-line argument to `sudo`.  The intended innermost level of quoting lies entirely on line 397 and occurs three times --- around ${PROXY_SETTING}, ${MASTER_IP}, and ${EXTRA_SANS}.  But the inner dquotes are not escaped properly.  What actually happens is that the outer level of quoting begins on line 379 and ends early, at the start of line 397.  That quoted string is concatenated with the expansion of ${PROXY_SETTING}, then " ~/kube/make-ca-cert.sh ", then the expansion of ${MASTER_IP}, then a space, then the expansion of ${EXTRA_SANS}, then another long string that begins at the end of line 397 and ends at the beginning of line 402.

Here is a demonstration of how a shorter example goes wrong:
mspreitz@ovespa003:~$ X="arg2; cat"
mspreitz@ovespa003:~$ ssh localhost "sudo /bin/bash -c 'echo arg1 "${X}" arg3' " 
arg1 arg2
cat: arg3: No such file or directory

MikeSpreitzer commented on cluster/ubuntu/util.sh:
```diff
@@ -127,17 +129,11 @@ function verify-cluster {
-  echo
-  echo "  http://${MASTER_IP}:8080"
-  echo
-
```
Why are these lines being deleted?

dalanlan replied:
It's a dup. ref #16838

MikeSpreitzer replied:
OK, thanks for the explanation.

MikeSpreitzer commented on cluster/ubuntu/util.sh:
```diff
@@ -147,24 +143,24 @@ function verify-master(){
         ((try_count=try_count+1))
         if [[ ${try_count} -gt ${max_try_count} ]]; then
-          printf "\nWarning: Process \"${daemon}\" failed to run on ${MASTER}, please check.\n"
+          echo -e "\nWarning: Process \"${daemon}\" failed to run on ${MASTER}, please check.\n"
```
By dropping the -n you have added a trailing blank line here.

dalanlan replied:
Yeah, that's expected

MikeSpreitzer commented on cluster/ubuntu/util.sh:
```diff
@@ -263,11 +265,11 @@ EOF
   setClusterInfo
-  KUBE_MASTER=$MASTER
-  KUBE_MASTER_IP=$MASTER_IP
+  KUBE_MASTER=${MASTER}
```
Note the lack of quoting around the variable references on the right hand side.  This is not consistent with other parts of this script, which attempt triply-nested quoting in order to quote the same variable references.

Is the quoting needed?  If not, then how about changing the script to never do it?
If you are unsure whether the quoting is needed, then you should quote every reference to the same variable.

MikeSpreitzer commented on cluster/ubuntu/util.sh:
```diff
@@ -294,7 +296,8 @@ function detect-nodes {
 
   if [[ -z "${KUBE_NODE_IP_ADDRESSES[@]}" ]]; then
-    echo "Could not detect Kubernetes node nodes. Make sure you've launched a cluster with 'kube-up.sh'" >&2
+    echo -e "Could not detect Kubernetes node nodes.\
```
You do not need the -e here.

MikeSpreitzer commented on cluster/ubuntu/util.sh:
```diff
@@ -305,8 +308,8 @@ function kube-up() {
   if [ ! -f "ubuntu/binaries/master/kube-apiserver" ]; then
-    echo "No local binaries for kube-up, downloading... "
-    "${KUBE_ROOT}/cluster/ubuntu/build.sh"
+    echo -n "No local binaries for kube-up, downloading..."
```
Why is -n added here?

MikeSpreitzer commented on cluster/ubuntu/util.sh:
```diff
@@ -346,99 +349,229 @@ function kube-up() {
+      cp ~/kube/init_scripts/* /etc/init.d/
+      
+      groupadd -f -r kube-cert
+      "${PROXY_SETTING}" ~/kube/make-ca-cert.sh "${MASTER_IP}" "${EXTRA_SANS}"
```
Botched triple nested quoting here too.

MikeSpreitzer commented on cluster/ubuntu/util.sh:
```diff
@@ -447,7 +580,8 @@ function kube-down {
 function prepare-push() {
   # Use local binaries for kube-push
-  if [[ "${KUBE_VERSION}" == "" ]]; then
+  if [[ -z ${KUBE_VERSION} ]]; then
```
You want to quote the variable reference here, in case (at least) it expands to the empty string.

MikeSpreitzer commented on cluster/ubuntu/util.sh:
```diff
@@ -456,14 +590,14 @@ function prepare-push() {
 # Update a kubernetes master with expected release
-function push-master {
+function push-master() {
   source "${KUBE_ROOT}/cluster/ubuntu/${KUBE_CONFIG_FILE-"config-default.sh"}"
```
I see two problems here.  First, it looks like nested quoting is intended here --- but is not done correctly.  BTW, the inner quoting is not needed.  Second, it looks like variable expansion with a default value is intended here, but it is also not done correctly --- the syntax for defaulting is :- not -.

Here is what `man bash` has to say about expansion with a default value:
```
       ${parameter:-word}
              Use Default Values.  If parameter is unset or null,  the  expan‐
              sion  of word is substituted.  Otherwise, the value of parameter
              is substituted.
```
All of config-default.sh is one word, in bash syntax.

zmerlynn commented:
BTW, I highly recommend `shellcheck`. It's nitpicky but basically lint for bash. I've started using it to evaluate new lines for PRs (manually checking the line numbers, it doesn't have an easy diff mode that I know of).

I'm leaving this up to @MikeSpreitzer for now. @MikeSpreitzer, if you LGTM it, I'll give it a once over and label it appropriately.

MikeSpreitzer commented:
@zmerlynn: you mean the one at shellcheck.net?

zmerlynn commented:
@MikeSpreitzer: Yes (http://hackage.haskell.org/package/ShellCheck)

MikeSpreitzer commented:
shellcheck.net made 77 comments, most pretty stupid (variables not assigned, lack of quotes around expansion of a variable that certainly expands to a word).  Wading through the rest now...

MikeSpreitzer commented on cluster/ubuntu/util.sh:
```diff
@@ -107,7 +109,7 @@ function trap-add {
   trap "${handler}" ${signal}
```
From shellchck.net: SC2064 (https://github.com/koalaman/shellcheck/wiki/SC2064) Use single quotes, otherwise this expands now rather than when signalled.

dalanlan replied:
I'm trying to use single quotes here, yet it complains that `handler` is a binding var. 
I'll dive down later.

MikeSpreitzer replied:
Oh!  Double quotes is correct here.  You want the trap to use the current (and local!) expansion of handler.

MikeSpreitzer commented on cluster/ubuntu/util.sh:
```diff
@@ -305,8 +308,8 @@ function kube-up() {
-    echo "No local binaries for kube-up, downloading... "
-    "${KUBE_ROOT}/cluster/ubuntu/build.sh"
+    echo -n "No local binaries for kube-up, downloading..."
+    ${KUBE_ROOT}/cluster/ubuntu/download-release.sh
```
From shellcheck.net: SC2086 (https://github.com/koalaman/shellcheck/wiki/SC2086) Double quote to prevent globbing and word splitting.

MikeSpreitzer commented on cluster/ubuntu/util.sh:
```diff
@@ -346,99 +349,229 @@ function kube-up() {
+      mkdir -p /opt/bin/
+      cp ~/kube/master/* /opt/bin/
+      service etcd start
+      FLANNEL_NET="${FLANNEL_NET}" ~/kube/reconfDocker.sh a
```
Same triply-nested quoting issue here.

zmerlynn commented:
Quoting is something I've started commenting on religiously, even if it seems stupid. Mainly because bash is stupid, and it's easier for the reader if it's done prescriptively than not. That actually matches our [style guide](http://kubernetes.io/v1.1/docs/devel/coding-conventions.html) as well.

zmerlynn commented:
That said, if it's old lines, feel free to omit or use your judgement.

MikeSpreitzer commented on cluster/ubuntu/util.sh:
```diff
@@ -515,21 +661,35 @@ function push-node() {
+  
+  local node_ip=${1}
   local ii=0
   local localexisting=false
```
The variable localexisting is un-referenced in this file.  Do we need to assign to it?

dalanlan replied:
It's a bug introduced by #16842. Should be `existing` instead. I' ll fix this.

MikeSpreitzer commented on cluster/ubuntu/util.sh:
```diff
@@ -515,21 +661,35 @@ function push-node() {
   local localexisting=false
 
   for i in ${nodes}; do
     if [[ "${roles[${ii}]}" == "i" && ${i#*@} == $node_ip ]]; then
```
I suggest quoting all the variable references in this line.

MikeSpreitzer commented on cluster/ubuntu/util.sh:
```diff
@@ -346,99 +349,229 @@ function kube-up() {
+    DNS:kubernetes.default.svc.cluster.local
+  )
+  
+  EXTRA_SANS=$(echo ${EXTRA_SANS[@]} | tr ' ' ',')
```
You do not have to quote the comma, it is not a shell metacharacter.

MikeSpreitzer commented:
OK, @dalanlan, I have finished my review of this comment.  Don't take my lack of flagging some un-quoted variable references as any contradictory to what @zmerlynn said.  Regarding rule 1 of https://google.github.io/styleguide/shell.xml?showone=Variable_expansion#Variable_expansion ("stay consistent with what you find"), I think that in this case what we found was not self-consistent so rule 1 is null and void.

MikeSpreitzer commented:
Or we could take the view that rule 1 means that the pre-existenting self-inconsistency is license for us to leave it self-inconsistent.

dalanlan commented:
I've tested it on shellcheck and addressed most of the comments (not for the `trap` one).
Update the code now in case @MikeSpreitzer might want a check (especially for the messy quote thing).

dalanlan commented on cluster/ubuntu/util.sh:
```diff
@@ -346,99 +347,230 @@ function kube-up() {
+      
+      groupadd -f -r kube-cert
+      \"${PROXY_SETTING}\" ~/kube/make-ca-cert.sh \"${MASTER_IP}\" \"${EXTRA_SANS}\"
+      exit 1
```
Oh my! It's used for testing:-/

dalanlan committed:
fix kube-up workflow

dalanlan commented:
ping @MikeSpreitzer

MikeSpreitzer commented on cluster/ubuntu/util.sh:
```diff
@@ -294,20 +294,21 @@ function detect-nodes {
-    echo "No local binaries for kube-up, downloading... "
-    "${KUBE_ROOT}/cluster/ubuntu/build.sh"
+  # downloading tarball release
+  if [[ -d "${KUBE_ROOT}/cluster/ubuntu/binaries" ]]; then
```
This is a logic change, and I do not understand the system well enough to say whether it is good or bad.

dalanlan replied:
The old one has potential bug (my bad) to bring up a cluster with an unmatched version due to the legacy file.

MikeSpreitzer commented on cluster/ubuntu/util.sh:
```diff
@@ -346,99 +347,229 @@ function kube-up() {
+              /etc/default/etcd
+      
+            rm -rf /infra*
+            rm -rf /srv/kubernetes
```
I don't know why this was added.

dalanlan replied:
A simple clean-up as well.

MikeSpreitzer replied:
I may not have been clear.  My question is why was the removal of /srv/kubernetes added?

dalanlan replied:
I believe kube-down should do all of the clean-up work. What's involved this directory is legacy files created by `make-ca-cert.sh`

MikeSpreitzer commented:
@dalanlan, @zmerlynn: I am not expert enough to evaluate the logic changes.  The quoting changes look good enough to me.  There are more nits we could pick, but I think they could be done later if anybody cares; this change is already uncomfortably large and long-in-progress, let's get it merged.  I am more interested in more substantial changes.

zmerlynn commented:
@MikeSpreitzer: That sounded like an LGTM from you, right? Looking.

zmerlynn commented:
LGTM. @dalanlan: In the future, I'd suggest smaller, bite sized PRs, or splitting the PR into smaller commits. This had a large combination of cleanup and some logic change, which makes it tricky to review (especially when it's outside the expertise of a lot of us). It's not a big deal, but it might decrease time-to-review. Sorry if this took a bit.

MikeSpreitzer commented:
Yes, LGTM from me.  Agree about smaller PRs.

dalanlan commented:
@zmerlynn  @MikeSpreitzer  Agreed. Thanks:)
